tls/s2n_cipher_suites.c

```diff
@@ -1155,17 +1173,191 @@ int s2n_set_cipher_as_client(struct s2n_connection *conn, uint8_t wire[S2N_TLS_C
+    /** 
+     * Tracks mutually supported cipher suite in the client index. This is needed to track tie-breakers
+     * in an equal-preference group.
+     */
+    int negotiated_client_index = count;
+    /* Same as negotiated_client_index, but is instead a server index. */
+    int negotiated_server_index = -1;
+    /** 
+     * The highest protocol version match index (also a server index). This is a fall back index if
+     * no other mutually-supported cipher suite is identified.
+     */
+    int highest_vers_match_index = -1;
+
+    for (uint32_t i = 0; i < cipher_preferences->count; i++) {
+        const struct s2n_cipher_suite* server_cipher = cipher_preferences->suites[i];
+        RESULT_ENSURE_REF(server_cipher);
+
+        /* Check if the cipher suite is an equal-preference grouping delimiter */
+        if (server_cipher == &s2n_equal_preference_group_start) {
+            in_group = true;
+            continue;
+        } 
+        if (server_cipher == &s2n_equal_preference_group_end) {
+            in_group = false;
+            /* Exiting a group and a negotiated cipher has already been found. */
+            if (negotiated_server_index != -1) {
+                *server_index = negotiated_server_index;
+                return S2N_RESULT_OK;
+            }
+            continue;
+        }
+
+        /* Cipher suite is NOT a delimiter */
+        int client_index = s2n_wire_index_of_cipher_suite(server_cipher->iana_value, wire, count, cipher_suite_len);
+
+        /* Client does not support this cipher. Skip. */
+        if (client_index < 0) {
+            continue;
+        }
+
+        /* Found a potential match. Validate that the cipher suite is suitable for this connection. */
+        struct s2n_cipher_suite* match = security_policy->cipher_preferences->suites[i];
+        if (!s2n_cipher_suite_match_is_valid(conn, match)) {
+            continue;
+        }
+
+        /* Don't immediately choose a cipher the connection shouldn't be able to support */
+        if (conn->actual_protocol_version < match->minimum_required_tls_version) {
+            if (highest_vers_match_index == -1) {
+                highest_vers_match_index = i;
+            }
+            continue;
+        }
+        if (in_group) {
+            /* Both client and server support a grouped-cipher */
+            if (client_index < negotiated_client_index) {
+                negotiated_client_index = client_index;
+                negotiated_server_index = i;
+            }
+        } else {
+            /* Else, the client and server both support a non-grouped cipher. */
+            negotiated_server_index = i;
+            break;
+        }
+    }
+
+    /* Found a match */
+    if (negotiated_server_index != -1) {
+        *server_index = negotiated_server_index;
+        return S2N_RESULT_OK;
+    }
+
+    /* Settle for a cipher with a higher required proto version, if it was set */
+    if (highest_vers_match_index != -1) {
```

Comment:
we don't always go through this check as we might return at line 1308

Reply:
Yes, that's correct behaviour. The return in line 1308 means we were able to identify a tie-breaker cipher suite

Reply:
but we still need to validate that the tie breaker is with a higher required protocol version (if it was set ), right ?

Reply:
I don't think so. In the current implementation, we choose the first valid cipher and return (https://github.com/aws/s2n-tls/pull/3488/files#diff-d0b6d9d296ee7b71a99c83ad1423142d1cc7d6b62a84ebfc31a9b0f633a4f323L1266). The `higher_vers_match` is saved as a fall-back cipher and is only used if could not find another valid cipher; we then fall back to the higher version as a last resort (https://github.com/aws/s2n-tls/pull/3488/files#diff-d0b6d9d296ee7b71a99c83ad1423142d1cc7d6b62a84ebfc31a9b0f633a4f323L1272)

Reply:
thanks